docs/proposals/karmada-operator/crd-http-source-proxy/README.md

```diff
@@ -0,0 +1,82 @@
+Currently, the Karmada operator's custom HTTP source CRD download strategy allows specifying an HTTP(S) URL as the source for the CRD tarball but does not support proxy configuration.
+This limitation prevents the operator from functioning correctly in such restricted environments. By adding support for specifying a proxy, we can ensure that the Karmada operator 
+is flexible and adaptable to diverse network configurations.
+
+## Goals
+
+- Enable proxy configuration for downloading CRD tarballs from custom HTTP/HTTPS sources.
+- Maintain backward compatibility by keeping the proxy field optional.
+- Ensure that the proxy configuration, if specified, is honored when downloading CRDs in both `Always` and `IfNotPresent` policies.
+
+## Proposal
+
+This proposal extends the `HTTPSource` type in the Karmada CRD specification to include an optional `Proxy` field. This field will specify the proxy server to use when downloading the CRD tarball.
+
+### API Changes
+
+Update the `HTTPSource` type as follows:
+
+```go
+// HTTPSource specifies how to download the CRD tarball via either HTTP or HTTPS protocol.
+type HTTPSource struct {
+    // URL specifies the URL of the CRD tarball resource.
+    URL string `json:"url,omitempty"`
+    
+    // Proxy specifies the HTTP/HTTPS proxy server to use when downloading the CRD tarball.
+    // The format should be a valid URL, e.g., "http://proxy.example.com:8080".
+    // +optional
+    Proxy *string `json:"proxy,omitempty"`
+}
+```
+
+### Behavior
+
+- If the `Proxy` field is set, the operator will configure the HTTP client to route requests through the specified proxy.
+- If the `Proxy` field is not set, the HTTP client will behave as it does today (i.e., direct connection or system-level proxy settings).
+- The proxy setting will apply only to the download of the CRD tarball. It will not affect other HTTP operations performed by the operator.
+
+### Caching Behavior
+
+This proposal does not alter the caching logic. The cache key for a given CRD tarball will continue to be derived from the URL alone. The proxy does not affect the identity of the downloaded content and therefore does not contribute to the cache key.
+
+### Design Considerations
+
+- **Extensibility**: The design allows for future support of authentication or SOCKS proxies if needed.
+
+### Alternatives Considered
+
+- Using environment variables to configure the proxy. This was rejected because it lacks the granularity and clarity of defining proxy settings on a per-resource basis within the CRD.
+
+### Implementation Plan
+
+- Extend the `HTTPSource` struct to include the `Proxy` field.
+- Update the CRD schema and validation logic.
+- Modify the download logic to honor the `Proxy` setting when set.
+
+### Test Plan
+
+- E2E tests in an environment requiring a proxy to ensure successful CRD downloads.
+- E2E tests to verify that the behavior remains unchanged when the `Proxy` field is omitted.
```

Comment:
Any idea to implement the E2E? We need to set up a proxy and http server, or just need a dummy proxy?

Reply:
> E2E tests to verify that the behavior remains unchanged when the `Proxy` field is omitted.

This is the legacy feature. Do we have E2E test for it?  @zhzhuang-zju Do you know?

Reply:
> > E2E tests to verify that the behavior remains unchanged when the `Proxy` field is omitted.
> 
> This is the legacy feature. Do we have E2E test for it? @zhzhuang-zju Do you know?

We have coverage for this in [E2E tests for the operator](https://github.com/karmada-io/karmada/blob/140f745e031e966b4df63f73f8573a6087881b58/hack/local-up-karmada-by-operator.sh#L84).